common/elasticsearch/bulk/bulk_index_request.go

```diff
@@ -0,0 +1,190 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package bulk
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+type BulkIndexRequest struct {
+	GenericBulkableRequest
+	index       string
+	typ         string
+	id          string
+	opType      string
+	version     *int64
+	versionType string // default is "internal"
+	doc         interface{}
+
+	source []string
+}
+
+type bulkIndexRequestCommand map[string]bulkIndexRequestCommandOp
+
+type bulkIndexRequestCommandOp struct {
+	Index       string `json:"_index,omitempty"`
+	ID          string `json:"_id,omitempty"`
+	Version     *int64 `json:"version,omitempty"`
+	VersionType string `json:"version_type,omitempty"`
+}
+
+// NewBulkIndexRequest returns a new BulkIndexRequest.
+// The operation type is "index" by default.
+func NewBulkIndexRequest() *BulkIndexRequest {
+	return &BulkIndexRequest{
+		opType: "index",
+	}
+}
+
+// Index specifies the Elasticsearch index to use for this index request.
+// If unspecified, the index set on the BulkService will be used.
+func (r *BulkIndexRequest) Index(index string) *BulkIndexRequest {
+	r.index = index
+	r.source = nil
+	return r
+}
+
+// Type specifies the Elasticsearch type to use for this index request.
+// If unspecified, the type set on the BulkService will be used.
+func (r *BulkIndexRequest) Type(typ string) *BulkIndexRequest {
+	r.typ = typ
+	r.source = nil
+	return r
+}
+
+// ID specifies the identifier of the document to index.
+func (r *BulkIndexRequest) ID(id string) *BulkIndexRequest {
+	r.id = id
+	r.source = nil
+	return r
+}
+
+// OpType specifies if this request should follow create-only or upsert
+// behavior. This follows the OpType of the standard document index API.
+// See https://www.elastic.co/guide/en/elasticsearch/reference/7.0/docs-index_.html#operation-type
+// for details.
+func (r *BulkIndexRequest) OpType(opType string) *BulkIndexRequest {
+	r.opType = opType
+	r.source = nil
+	return r
+}
+
+// Version indicates the version of the document as part of an optimistic
+// concurrency model.
+func (r *BulkIndexRequest) Version(version int64) *BulkIndexRequest {
+	v := version
+	r.version = &v
```

Comment:
```suggestion
	r.version = &version
```
it's already a copy (value arg), no need to make another.

Reply:
done